File: data/code_pairs/cpp/2.cpp
#include <iostream>
#include <complex>

void CLAPMT(bool forwrd, int m, int n, std::complex<float>* X, int ldx, int* K) {
    if (n <= 1) {
        return;
    }

    for (int i = 0; i < n; ++i) {
        K[i] = -K[i];
    }

    if (forwrd) {
        for (int i = 0; i < n; ++i) {
            if (K[i] > 0) continue;

            int j = i;
            K[j] = -K[j];
            int in = K[j] - 1;

            while (true) {
                if (K[in] > 0) break;

                for (int ii = 0; ii < m; ++ii) {
                    std::complex<float> temp = X[ii + j * ldx];
                    X[ii + j * ldx] = X[ii + in * ldx];
                    X[ii + in * ldx] = temp;
                }

                K[in] = -K[in];
                j = in;
                in = K[in] - 1;
            }
        }
    } else {
        for (int i = 0; i < n; ++i) {
            if (K[i] > 0) continue;

            K[i] = -K[i];
            int j = K[i] - 1;

            while (true) {
                if (j == i) break;

                for (int ii = 0; ii < m; ++ii) {
                    std::complex<float> temp = X[ii + i * ldx];
                    X[ii + i * ldx] = X[ii + j * ldx];
                    X[ii + j * ldx] = temp;
                }

                K[j] = -K[j];
                j = K[j] - 1;
            }
        }
    }
}

File: data/code_pairs/fortran/2.f90
SUBROUTINE CLAPMT(FORWRD, M, N, X, LDX, K)
    LOGICAL            FORWRD
    INTEGER            LDX, M, N
    INTEGER            K(*)
    COMPLEX            X(LDX,*)

    INTEGER            I, II, J, IN
    COMPLEX            TEMP

    IF (N.LE.1) RETURN

    DO I = 1, N
        K(I) = -K(I)
    END DO

    IF (FORWRD) THEN
        DO I = 1, N
            IF (K(I) > 0) GO TO 40

            J = I
            K(J) = -K(J)
            IN = K(J)

            DO WHILE (K(IN) <= 0)
                DO II = 1, M
                    TEMP = X(II, J)
                    X(II, J) = X(II, IN)
                    X(II, IN) = TEMP
                END DO
                K(IN) = -K(IN)
                J = IN
                IN = K(IN)
            END DO
40          CONTINUE
        END DO
    ELSE
        DO I = 1, N
            IF (K(I) > 0) GO TO 100

            K(I) = -K(I)
            J = K(I)
            DO WHILE (J /= I)
                DO II = 1, M
                    TEMP = X(II, I)
                    X(II, I) = X(II, J)
                    X(II, J) = TEMP
                END DO
                K(J) = -K(J)
                J = K(J)
            END DO
100     CONTINUE
        END DO
    END IF
END SUBROUTINE CLAPMT